### PR TITLE
adding a method to transform ResultSet into a usable format for Weka …

### PR DESCRIPTION
**Pull request by Melviyn**
adding a method to transform ResultSet into a usable format for Weka (Instances) and JUnit test for this method

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -77,7 +77,9 @@
         <dependency>
             <groupId>org.orbisgis</groupId>
             <artifactId>h2spatial</artifactId>
-            <version>1.2.4-SNAPSHOT</version>
+            <version>1.1.1-SNAPSHOT</version>
```
Comment (ebocher) [inline pom.xml]: Why this version instead the last one ?

File: src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java [hunk 2]
```diff
@@ -50,9 +56,78 @@ public void tearDownStatement() throws Exception {
...
+    public void testCreateInstancesFromDB() throws SQLException, IOException, Exception {
         Statement stat = connection.createStatement();
         //Create the tables
+        String sql = "CREATE TABLE COMPANY " +
+                      "(ID INT PRIMARY KEY     NOT NULL," +
+                      " NAME           TEXT    NOT NULL, " +
+                      " AGE            INT     NOT NULL, " +
+                      " ADDRESS        VARCHAR(50), " +
+                      " SALARY         REAL)";
+        stat.executeUpdate(sql);
```
Comment (ebocher) [inline src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java:111]: use execute

File: src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java [hunk 3]
```diff
@@ -50,9 +56,78 @@ public void tearDownStatement() throws Exception {
...
+        
+        //===========ResultSet contains something==============
+        ResultSet res = stat.executeQuery(sql);
+
+        assertTrue("failure - ResultSet have nothing", res.next());
+        res.close();
+        
+        //===========Instances contains something==============
+        res = stat.executeQuery(sql);
+        String pathname = "/home/mlgall/Documents/RandomForest/data/RandomForest.model";
```
Comment (ebocher) [inline src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java]: Please use target folder to store temporary resource

**Comment by ebocher** [comment]
Please do not merge.

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        
+        connection = SpatialH2UT.openSpatialDataBase(DB_NAME);
+        if(connection == null){
+           connection = SpatialH2UT.createSpatialDataBase(DB_NAME);
+        }
+        stat = connection.createStatement();
+    }
+    
+    /**
+     * Initialize the Classifier RandomForest with a RandomForest.model file
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: With a fixed model build for the MApUCE project

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+     * Work for simple data type
+     * @param res ResultSet got after an SQL request
+     * @throws Exception 
+     */
+    public Instances resultSetToInstances(ResultSet rs) throws Exception{
+
+        Instances inst = InstanceQuery.retrieveInstances(new InstanceQuery(),rs);
+        
+        inst.insertAttributeAt(attClass, inst.numAttributes());
+        //inst.replaceAttributeAt(attClass, inst.numAttributes()-1);
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: Why this line ?

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        return inst;
+    }
+    
+  
+    /**
+     * Execute request via a connection to the database 
+     * @param query query like "SELECT * FROM the_table"
+     * @return ResultSet of the request
+     */
+    public ResultSet executeQuery(String query,String nameColPk) throws SQLException{
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: What about if the resulset doesn't contain the nameColPk ?

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            Instance newInst = dataToClassify.instance(i); 
+            
+            double pk = newInst.value(this.colPk);
+            double predNB = model.classifyInstance(newInst);
+            
+            String predString = dataToClassify.classAttribute().value((int)predNB);
+            newInst.setClassValue(predString);             
+            ret.put(pk,predString);
+            if(ret.size() >= BATCH_MAX_SIZE){
+                tableResult("tableResultTest",ret);
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: Why this name by default ?

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            st.setString(2, ret.get(db));
+            st.addBatch();
+        }
+          
+        st.executeBatch();
+        
+        String sql = "SELECT * FROM "+nameTableResult;
+        ResultSet res = stat.executeQuery(sql);
+        while(res.next()){
+           System.out.println(res.getDouble(1)+": "+res.getString(2));
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: System. out ????

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+            
+          }
+          if(ret.size() < BATCH_MAX_SIZE){
+              tableResult("tableResultTest",ret);
+                ret.clear();
+          }
+
+        }
+        
+        return ret;
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: Why are u returning a HashMap ? the data has been insert in the table ?

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+     * Method for update the class of each line in database
+     */
+    private void tableResult(String nameTableResult,HashMap<Double,String> ret) throws SQLException{
+        
+        //création table avec deux col id - class
+        PreparedStatement st = null;
+        //prepareStatement
+        
+        String createTableSQL =  "CREATE TABLE IF NOT EXISTS "+nameTableResult+"("
+				+ "ID REAL NOT NULL, "
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: INT ?

File: src/main/resources/org/orbisgis/mapuce/randomForest/ResultSet.arff~ [hunk 11]
```diff
@@ -0,0 +1,219 @@
+% 1. Title: Iris Plants Database
```
Comment (ebocher) [inline src/main/resources/org/orbisgis/mapuce/randomForest/ResultSet.arff~]: duplicate file ?

File: src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java [hunk 12]
```diff
@@ -5,145 +5,131 @@
...
+import java.util.HashMap;
 import org.junit.Test;
 import static org.junit.Assert.*;
 import weka.core.Instances;
-import weka.experiment.InstanceQuery;
 
 /**
  *
- * @author ebocher
+ * @author ebocher Melvin Le Gall
```
Comment (ebocher) [inline src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java]: @author ebocher
@author Melvin Le Gall

File: src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java [hunk 13]
```diff
@@ -5,145 +5,131 @@
...
         stat.execute(sql);
         
-        sql = "CREATE TABLE COMPANY " +
-                "(ID INT PRIMARY KEY     NOT NULL," +
-                " NAME           VARCHAR(50)   NOT NULL, " +
-                " AGE            INT     NOT NULL, " +
-                " ADDRESS        VARCHAR(50), " +
-                " SALARY         REAL)";
+        sql = "CREATE TABLE IRIS " +
+                "(sepallength    REAL     NOT NULL," +
```
Comment (ebocher) [inline src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java]: REAL ?

File: src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java [hunk 14]
```diff
@@ -5,145 +5,131 @@
...
-        
+        /*
         //===========Compare each value for adress============
         res.close();
         res = stat.executeQuery(sql);
         //Compare value of each adress
         int j=0;
         while (res.next()) {
-            assertSame("should be same", res.getString(4), inst.get(j).stringValue(3));
+            //assertSame("should be same", res.getString(4), inst.get(j).stringValue(3));
```
Comment (ebocher) [inline src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java]: All results must be tested.

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 15]
```diff
@@ -56,58 +55,61 @@
...
      */
-    public Instances resultSetToInstances(ResultSet rs) throws Exception{
+    public Instances resultSetToInstances(ResultSet rs,String nameColIndex) throws Exception{
 
+        try{
+           nameTypeIndex = rs.getMetaData().getColumnTypeName(rs.findColumn(nameColIndex));
+           this.colIndex = rs.findColumn(nameColIndex)-1; 
+        }
+        catch(SQLException e){
+            LoggerFactory.getLogger(ClassifyData.class).error(e.getMessage());
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: use
private Logger LOGGER = LoggerFactory.getLogger(ClassifyData.class);

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 16]
```diff
@@ -116,28 +118,14 @@ public Instances resultSetToInstances(ResultSet rs) throws Exception{
...
-    }
-       
     /**
      * Classify data after executeQuery
-     * @return ArrayList with each prediction line by line
+     * @param nameTableResult table wich is stored result
      * @throws Exception 
      */
-    public HashMap<Double, String> classifyData(String nameTableResult) throws Exception{
+    public void classify(String nameTableResult) throws Exception{
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java:123]: Use org.h2gis.utilities.TableLocation to return a well formated table name

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 17]
```diff
@@ -56,58 +55,61 @@
...
+     * @param rs ResultSet got after an SQL request
+     * @param nameColIndex Name of the collumn used to trace each line
      * @throws Exception 
      */
-    public Instances resultSetToInstances(ResultSet rs) throws Exception{
+    public Instances resultSetToInstances(ResultSet rs,String nameColIndex) throws Exception{
 
+        try{
+           nameTypeIndex = rs.getMetaData().getColumnTypeName(rs.findColumn(nameColIndex));
+           this.colIndex = rs.findColumn(nameColIndex)-1;
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: rs.findColumn two times ?

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 18]
```diff
@@ -147,83 +135,99 @@ public ResultSet executeQuery(String query,String nameColPk) throws SQLException
...
-        //prepareStatement
         
         String createTableSQL =  "CREATE TABLE IF NOT EXISTS "+nameTableResult+"("
-				+ "ID REAL NOT NULL, "
+				+ "INDEX "+nameTypeIndex+ " NOT NULL, "
 				+ "CLASS VARCHAR(50) NOT NULL, "
 				+ ")";
         
         st = connection.prepareStatement(createTableSQL);
         st.execute();
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: st.execute(createTableSQL) is enough.

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 19]
```diff
@@ -116,28 +118,14 @@ public Instances resultSetToInstances(ResultSet rs) throws Exception{
...
+     * @param nameTableResult table wich is stored result
      * @throws Exception 
      */
-    public HashMap<Double, String> classifyData(String nameTableResult) throws Exception{
+    public void classify(String nameTableResult) throws Exception{
     
-        HashMap<Double,String> ret = new HashMap<>();
+        HashMap<Object,String> ret = new HashMap<>();
         
         this.nameTableResult = nameTableResult;
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: Not used anymore?

File: src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java [hunk 20]
```diff
@@ -56,58 +55,61 @@
...
-        connection = SpatialH2UT.openSpatialDataBase(DB_NAME);
-        if(connection == null){
-           connection = SpatialH2UT.createSpatialDataBase(DB_NAME);
-        }
+        model = (RandomForest) obj[0];
+        Instances inst = (Instances)obj[1];
+        attClass = inst.classAttribute();
+
+        connection = conn;
         stat = connection.createStatement();
```
Comment (ebocher) [inline src/main/java/org/orbisgis/mapuce/randomForest/ClassifyData.java]: Useful ?

File: src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java [hunk 21]
```diff
@@ -6,131 +6,148 @@
...
+    private static Connection connection;
+    private static final String DB_NAME = "RandomForestTest";
+    private Statement stat;
+
+    
+    @BeforeClass
+    public static void tearUp() throws Exception {
+        // Keep a connection alive to not close the DataBase on each unit test
+        connection = SpatialH2UT.openSpatialDataBase(DB_NAME);
+        if(connection == null){
```
Comment (ebocher) [inline src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java]: Don't understand this test ?

File: src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java [hunk 22]
```diff
@@ -6,131 +6,148 @@
...
+        stat = connection.createStatement();
+    }
+
+    @After
+    public void tearDownStatement() throws Exception {
+        stat.close();
+    }
+    
+    @Test
+    public void testCreateModel() throws SQLException, IOException, Exception {
```
Comment (ebocher) [inline src/test/java/org/orbisgis/mapuce/randomForest/RandomForestTest.java:59]: There is not test ?

**Comment by ebocher** [comment]
Is this PR ready to merge ?

**Comment by Melviyn** [comment]
Yes I have nothing to modify

**Pull request merged**